packages/meditrak-server/src/database/SurveyResponseChangeHandler/OutdatedStatusUpdater.js

```diff
@@ -0,0 +1,261 @@
+ */
+
+import assert from 'assert';
+import groupBy from 'lodash.groupby';
+import keyBy from 'lodash.keyby';
+import orderBy from 'lodash.orderby';
+import moment from 'moment';
+
+import { isMarkedChange } from '@tupaia/database';
+import { getUniqueEntries, getUniqueObjects, haveSameFields, max, min } from '@tupaia/utils';
+import { PERIOD_GRANULARITIES, PERIOD_GRANULARITY_TO_MOMENT_UNIT } from '../models/Survey';
+
+/**
+ * Survey responses use an `outdated` flag to indicate whether their values should be used
+ * for analytics building. This flag mainly depends on the survey response data time and on the
+ * `period_granularity` of its parent survey.
+ *
+ * * For surveys with no `period_granularity`: no responses are outdated, or in
+ * other words they all contribute towards analytics building
+ *
+ * * For  surveys with a `period_granularity` ("periodic"): only one response per
+ * survey/entity/period combination will contribute towards analytics building;
+ * the rest should be marked as outdated. This one response is selected on the basis of most
+ * recent submission time
+ *
+ * Here, we refer to the survey/entity/period combo as "response dimension combo".
+ * Example: Basic Clinic Survey/Tonga/July 2021
+ *
+ * @typedef { surveyId, entityId, startDate, endDate } ResponseDimensionCombo
+ */
+
+const DATE_FORMAT = 'YYYY-MM-DD';
+const DATETIME_FORMAT = `${DATE_FORMAT} HH:mm:ss`;
+
+/**
+ * Cache of period ranges keyed by requested period granularity first, then date
+ *
+ * ```js
+ * {
+ *   yearly: {
+ *     "2021-05-05": { startDate: '2021-01-01', endDate: '2021-12-31' }
+ *   }
+ *   monthly: {
+ *     "2021-07-06": { startDate: '2021-07-01', endDate: '2021-07-31' }
+ *   }
+ * }
+ * ```
+ */
+const periodRangeCache = Object.fromEntries(
+  Object.values(PERIOD_GRANULARITIES).map(granularity => [granularity, {}]),
+);
+
+/**
+ * @param {Date|string} datetime
+ * @returns {string}
+ */
+const extractDateString = datetime => {
+  const datetimeString = typeof datetime === 'string' ? datetime : datetime.toISOString();
+  return datetimeString.substring(0, DATE_FORMAT.length);
+};
+
+/**
+ */
+const getPeriodRange = (periodGranularity, datetime) => {
+  // None of the existing granularities depends on time, so we can just use the date part
+  // to cache/retrieve date ranges
+  const date = extractDateString(datetime);
+  const existingRange = periodRangeCache?.[periodGranularity]?.[date];
+  if (existingRange) {
+    return existingRange;
+  }
+
+  const momentUnit = PERIOD_GRANULARITY_TO_MOMENT_UNIT[periodGranularity];
+  const range = {
+    startDate: moment(datetime).startOf(momentUnit).format(DATETIME_FORMAT),
+    endDate: moment(datetime).endOf(momentUnit).format(DATETIME_FORMAT),
+  };
+  periodRangeCache[periodGranularity][date] = range;
+
+  return range;
+};
+
+const getIdOfMostRecentResponse = surveyResponses => {
+  assert.notStrictEqual(surveyResponses.length, 0);
```

Comment:
Nice, I haven't seen anyone use assertions using this library before but it certainly makes things tidy

Reply:
Yeap I wanted to comment on that! Happy that you like it. The idea is:

If it's a user (end user or dev) related error, throw an error. If it's just a precondition checking, use an assertion. So what this conveys is "the length of `surveyResponses` should really be <> 0 and supposedly the existing code doesn't allow that. But if for whatever reason this is not the case, let the devs know asap"

I find function preconditions and postconditions [Eiffel - Design By Contract](https://www.eiffel.org/doc/eiffel/ET-_Design_by_Contract_%28tm%29%2C_Assertions_and_Exceptions) an interesting mental pattern. Many mode modern languages haven't seem to picked up on that (perhaps it's too verbose?), but checking preconditions with one liner assertions seems like a good balance to me!

Reply:
Yeah that's a great rule, love it :-)